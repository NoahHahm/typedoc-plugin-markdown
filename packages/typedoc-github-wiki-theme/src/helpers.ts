export function parseUrl(url: string) {
  return encodeURI('../wiki/' + url.replace('.md', ''));
}
