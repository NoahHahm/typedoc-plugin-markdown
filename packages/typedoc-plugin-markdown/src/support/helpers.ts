import {
  ContainerReflection,
  DeclarationReflection,
  ProjectReflection,
  ReflectionKind,
  SignatureReflection,
} from 'typedoc';
import { backTicks } from './els';

/**
 * Determines if current signature is a constructor
 */
export const isConstructor = (
  signature: DeclarationReflection | SignatureReflection,
) => {
  return signature.parent?.kindOf(ReflectionKind.Constructor);
};

export function getIndexHeadingLevel(
  reflection: DeclarationReflection | ProjectReflection,
) {
  if (
    reflection.kindOf([
      ReflectionKind.Project,
      ReflectionKind.Module,
      ReflectionKind.Namespace,
    ]) ||
    reflection.hasOwnDocument
  ) {
    return 2;
  }
  return 4;
}

export function getGroupHeadingLevel(container: ContainerReflection) {
  return container.hasOwnDocument ? 2 : 4;
}

export function getReflectionHeadingLevel(
  reflection: DeclarationReflection | SignatureReflection,
) {
  if (reflection.hasOwnDocument) {
    return 1;
  }
  if (reflection.kindOf(ReflectionKind.Constructor)) {
    return reflection.parent?.hasOwnDocument ? 2 : 4;
  }
  return reflection.parent?.hasOwnDocument ? 3 : 5;
}

export function getMemberHeadingLevel(
  reflection: DeclarationReflection | SignatureReflection,
) {
  return getReflectionHeadingLevel(reflection) + 1;
}

export function getReflectionTitle(reflection: DeclarationReflection) {
  const md: string[] = [];

  md.push(reflection.name);

  if (reflection.typeParameters) {
    const typeParameters = reflection.typeParameters
      .map((typeParameter) => typeParameter.name)
      .join(', ');
    md.push(`\\<${typeParameters}\\>`);
  }

  return md.join('');
}

export function getFlags(reflection: DeclarationReflection) {
  if (reflection.flags?.length > 0 && !reflection.flags.isRest) {
    return reflection.flags.map((flag) => backTicks(flag)).join(' ');
  }
  return null;
}
