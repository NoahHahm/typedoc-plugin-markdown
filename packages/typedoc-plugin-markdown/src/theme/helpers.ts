/**
 *
 * A set of helpers that parses TypeDoc models to be consumed by theme resources.
 *
 * @module
 */

import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionKind,
} from 'typedoc';
import { backTicks } from '../support/elements';
import { escapeChars } from '../support/utils';

export function getDeclarationType(declaration: DeclarationReflection) {
  if (declaration.getSignature) {
    return declaration.getSignature.type;
  }
  if (declaration.setSignature) {
    return declaration.setSignature.type;
  }
  return declaration.type;
}

export function getProjectDisplayName(
  project: ProjectReflection,
  includeVersion: boolean,
): string {
  const version =
    includeVersion && project.packageVersion
      ? ` - v${project.packageVersion}`
      : '';
  return `${project.name}${version ? version : ''}`;
}

export function hasIndex(
  reflection: DeclarationReflection | ProjectReflection,
) {
  return reflection.groups?.some((group) => group.allChildrenHaveOwnDocument());
}

export function hasTOC(reflection: DeclarationReflection | ProjectReflection) {
  return (
    reflection.kindOf([
      ReflectionKind.Project,
      ReflectionKind.Module,
      ReflectionKind.Namespace,
    ]) &&
    reflection.groups?.some((group) => !group.allChildrenHaveOwnDocument())
  );
}

export function isGroupKind(reflection: DeclarationReflection) {
  return reflection.kindOf([
    ReflectionKind.Class,
    ReflectionKind.Interface,
    ReflectionKind.Enum,
    ReflectionKind.Function,
    ReflectionKind.Variable,
    ReflectionKind.TypeAlias,
  ]);
}

export function getModifier(reflection: DeclarationReflection) {
  if (reflection.flags.isAbstract) {
    return 'abstract';
  }
  if (reflection.flags.isPrivate) {
    return 'private';
  }
  if (reflection.flags.isReadonly) {
    return 'readonly';
  }
  if (reflection.flags.isStatic) {
    return 'static';
  }
  if (reflection.getSignature) {
    return 'get';
  }
  if (reflection.setSignature) {
    return 'set';
  }
  return null;
}

export const KEYWORD_MAP = {
  [ReflectionKind.Class]: 'class',
  [ReflectionKind.Interface]: 'interface',
  [ReflectionKind.Enum]: 'enum',
  [ReflectionKind.TypeAlias]: 'type',
};

export function getMemberTitle(reflection: DeclarationReflection) {
  const md: string[] = [];

  const name: string[] = [];

  const flagsToInclude = ['Abstract'];

  const flags = reflection.flags.filter((flag) =>
    flagsToInclude.includes(flag),
  );

  if (flags.length) {
    name.push(flags.map((flag) => backTicks(flag.toLowerCase())).join(' '));
    name.push(' ');
  }

  name.push(`${escapeChars(reflection.name)}`);

  if (reflection.typeParameters) {
    const typeParameters = reflection.typeParameters
      .map((typeParameter) => typeParameter.name)
      .join(', ');
    name.push(`${backTicks(`<${typeParameters}>`)}`);
  }

  md.push(name.join(''));

  return md.join(': ');
}
