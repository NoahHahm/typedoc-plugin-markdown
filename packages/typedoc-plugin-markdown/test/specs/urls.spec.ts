import {
  FixtureOutputDir,
  FixtureOutputFileStrategy,
} from '../__utils__/config';
import { expectUrlsToEqual } from '../__utils__/helpers';

describe(`Urls`, () => {
  test(`should gets Urls for multiple entry points`, () => {
    expectUrlsToEqual(FixtureOutputDir.Groups, [
      FixtureOutputFileStrategy.Members,
      FixtureOutputFileStrategy.Modules,
    ]);
  });

  test(`should gets Urls for single entry points`, () => {
    expectUrlsToEqual(FixtureOutputDir.Reflections, [
      FixtureOutputFileStrategy.Members,
      FixtureOutputFileStrategy.Modules,
    ]);
  });

  test(`should gets Urls for packages entry points`, () => {
    expectUrlsToEqual(FixtureOutputDir.Packages, [
      FixtureOutputFileStrategy.Members,
      FixtureOutputFileStrategy.Modules,
    ]);
  });
});
