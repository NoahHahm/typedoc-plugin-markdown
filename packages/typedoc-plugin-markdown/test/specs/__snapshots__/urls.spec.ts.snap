// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Urls should gets Urls for multiple entry points: outputFileStrategy: members 1`] = `
[
  "42-module-name/README.md",
  "<internal>/README.md",
  "@scope_name/README.md",
  "README.md",
  "basic/README.md",
  "basic/classes/ClassA.md",
  "basic/classes/ClassB.md",
  "basic/enumerations/EnumA.md",
  "basic/enumerations/EnumB.md",
  "basic/interfaces/InterfaceA.md",
  "basic/interfaces/InterfaceB.md",
  "basic/type-aliases/TypeA.md",
  "basic/type-aliases/TypeB.md",
  "has-categories/README.md",
  "has-categories/enumerations/CategoryAEnum1.md",
  "has-categories/enumerations/CategoryAEnum2.md",
  "has-categories/enumerations/CategoryBEnum1.md",
  "has-categories/enumerations/CategoryBEnum2.md",
  "has-categories/enumerations/UnCategorizedEnum.md",
  "has-categories/interfaces/CategoryAInterface1.md",
  "has-categories/interfaces/CategoryAInterface2.md",
  "has-categories/interfaces/CategoryBInterface1.md",
  "has-categories/interfaces/CategoryBInterface2.md",
  "has-categories/interfaces/UnCategorizedInterace.md",
  "has-namespaces/README.md",
  "has-namespaces/classes/ClassB.md",
  "has-namespaces/classes/Class_A_.md",
  "has-namespaces/enumerations/EnumB.md",
  "has-namespaces/enumerations/Enum_A_.md",
  "has-namespaces/functions/functionB.md",
  "has-namespaces/functions/function_A_.md",
  "has-namespaces/interfaces/InterfaceB.md",
  "has-namespaces/interfaces/Interface_A_.md",
  "has-namespaces/namespaces/NamespaceB/README.md",
  "has-namespaces/namespaces/Namespace_A_/README.md",
  "has-namespaces/namespaces/Namespace_A_/interfaces/NamespaceInterface.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace/README.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace/interfaces/NamespaceInterface.md",
  "has-namespaces/type-aliases/TypeB.md",
  "has-namespaces/type-aliases/Type_A_.md",
  "has-namespaces/variables/variableB.md",
  "has-namespaces/variables/variable_A_.md",
  "has-references/README.md",
  "has-references/functions/defaultFunction.md",
  "index/README.md",
  "modules.md",
]
`;

exports[`Urls should gets Urls for multiple entry points: outputFileStrategy: members 2`] = `
[
  "42-module-name/index.md",
  "<internal>/index.md",
  "@scope_name/index.md",
  "basic/classes/ClassA.md",
  "basic/classes/ClassB.md",
  "basic/enumerations/EnumA.md",
  "basic/enumerations/EnumB.md",
  "basic/index.md",
  "basic/interfaces/InterfaceA.md",
  "basic/interfaces/InterfaceB.md",
  "has-categories/enumerations/CategoryAEnum1.md",
  "has-categories/enumerations/CategoryAEnum2.md",
  "has-categories/enumerations/CategoryBEnum1.md",
  "has-categories/enumerations/CategoryBEnum2.md",
  "has-categories/enumerations/UnCategorizedEnum.md",
  "has-categories/index.md",
  "has-categories/interfaces/CategoryAInterface1.md",
  "has-categories/interfaces/CategoryAInterface2.md",
  "has-categories/interfaces/CategoryBInterface1.md",
  "has-categories/interfaces/CategoryBInterface2.md",
  "has-categories/interfaces/UnCategorizedInterace.md",
  "has-namespaces/classes/ClassB.md",
  "has-namespaces/classes/Class_A_.md",
  "has-namespaces/enumerations/EnumB.md",
  "has-namespaces/enumerations/Enum_A_.md",
  "has-namespaces/index.md",
  "has-namespaces/interfaces/InterfaceB.md",
  "has-namespaces/interfaces/Interface_A_.md",
  "has-namespaces/namespaces/NamespaceB/index.md",
  "has-namespaces/namespaces/Namespace_A_/index.md",
  "has-namespaces/namespaces/Namespace_A_/interfaces/NamespaceInterface.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace/index.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace/interfaces/NamespaceInterface.md",
  "has-references/index.md",
  "index.md",
  "module.index/index.md",
]
`;

exports[`Urls should gets Urls for multiple entry points: outputFileStrategy: modules 1`] = `
[
  "42-module-name.md",
  "<internal>.md",
  "@scope_name.md",
  "README.md",
  "basic.md",
  "has-categories.md",
  "has-namespaces/README.md",
  "has-namespaces/namespaces/NamespaceB.md",
  "has-namespaces/namespaces/Namespace_A_/README.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace.md",
  "has-references.md",
  "index.md",
  "modules.md",
]
`;

exports[`Urls should gets Urls for multiple entry points: outputFileStrategy: modules 2`] = `
[
  "42-module-name.md",
  "<internal>.md",
  "@scope_name.md",
  "basic.md",
  "has-categories.md",
  "has-namespaces/index.md",
  "has-namespaces/namespaces/NamespaceB.md",
  "has-namespaces/namespaces/Namespace_A_/index.md",
  "has-namespaces/namespaces/Namespace_A_/namespaces/NestedNamespace.md",
  "has-references.md",
  "index.md",
  "module.index.md",
]
`;

exports[`Urls should gets Urls for packages entry points: outputFileStrategy: members 1`] = `
[
  "README.md",
  "modules.md",
  "package-1/README.md",
  "package-1/interfaces/PackageInterface.md",
  "package-1/modules.md",
  "package-2/README.md",
  "package-2/interfaces/PackageInterface.md",
  "package-2/modules.md",
]
`;

exports[`Urls should gets Urls for packages entry points: outputFileStrategy: members 2`] = `
[
  "README.md",
  "package-1/README.md",
  "package-1/interfaces/PackageInterface.md",
  "package-1/modules.md",
  "package-2/README.md",
  "package-2/interfaces/PackageInterface.md",
  "package-2/modules.md",
]
`;

exports[`Urls should gets Urls for packages entry points: outputFileStrategy: modules 1`] = `
[
  "README.md",
  "modules.md",
  "package-1/README.md",
  "package-1/modules.md",
  "package-2/README.md",
  "package-2/modules.md",
]
`;

exports[`Urls should gets Urls for packages entry points: outputFileStrategy: modules 2`] = `
[
  "README.md",
  "package-1/README.md",
  "package-1/modules.md",
  "package-2/README.md",
  "package-2/modules.md",
]
`;

exports[`Urls should gets Urls for single entry points: outputFileStrategy: members 1`] = `
[
  "README.md",
  "classes/AbstractClass.md",
  "classes/ClassWithAccessors.md",
  "classes/ClassWithConstructorOverloads.md",
  "classes/ClassWithModifiers.md",
  "classes/ClassWithTypeParameters.md",
  "classes/DerivedClassA.md",
  "classes/DerivedClassB.md",
  "enumerations/BasicEnum.md",
  "enumerations/EnumWithValues.md",
  "exports.md",
  "functions/basicFunction.md",
  "functions/functionReturningAFunction.md",
  "functions/functionReturningAPromise.md",
  "functions/functionReturningAnObject.md",
  "functions/functionWithDefaultParameters.md",
  "functions/functionWithMultipleSignatures.md",
  "functions/functionWithNamedParams.md",
  "functions/functionWithNestedParameters.md",
  "functions/functionWithOptionalParameters.md",
  "functions/functionWithTypeParameters.md",
  "interfaces/BasicInterface.md",
  "interfaces/ExtendedInterface.md",
  "interfaces/IndexableInterface.md",
  "interfaces/InterfaceWithTypeParameters.md",
  "type-aliases/ArrayOfStuff.md",
  "type-aliases/ArrayType.md",
  "type-aliases/ConditionalType.md",
  "type-aliases/ExternalReferenceType.md",
  "type-aliases/FunctionType.md",
  "type-aliases/IndexAccessType.md",
  "type-aliases/IntersectionType.md",
  "type-aliases/LiteralType.md",
  "type-aliases/PartialMappedType.md",
  "type-aliases/PrimitiveType.md",
  "type-aliases/ReadonlyMappedType.md",
  "type-aliases/StringLiteralType.md",
  "type-aliases/TupleType.md",
  "type-aliases/TypeWithExternalSymbolLinkMapping.md",
  "type-aliases/TypeWithTypeParams.md",
  "type-aliases/UnionType.md",
  "variables/objectLiteralVariable.md",
  "variables/stringVariable.md",
  "variables/typeOperatorVariable.md",
]
`;

exports[`Urls should gets Urls for single entry points: outputFileStrategy: members 2`] = `
[
  "classes/AbstractClass.md",
  "classes/ClassWithAccessors.md",
  "classes/ClassWithConstructorOverloads.md",
  "classes/ClassWithModifiers.md",
  "classes/ClassWithTypeParameters.md",
  "classes/DerivedClassA.md",
  "classes/DerivedClassB.md",
  "enumerations/BasicEnum.md",
  "enumerations/EnumWithValues.md",
  "functions/basicFunction.md",
  "functions/functionReturningAFunction.md",
  "functions/functionReturningAPromise.md",
  "functions/functionReturningAnObject.md",
  "functions/functionWithDefaultParameters.md",
  "functions/functionWithMultipleSignatures.md",
  "functions/functionWithNamedParams.md",
  "functions/functionWithNestedParameters.md",
  "functions/functionWithOptionalParameters.md",
  "functions/functionWithTypeParameters.md",
  "index.md",
  "interfaces/BasicInterface.md",
  "interfaces/ExtendedInterface.md",
  "interfaces/IndexableInterface.md",
  "interfaces/InterfaceWithTypeParameters.md",
  "type-aliases/ArrayOfStuff.md",
  "type-aliases/ArrayType.md",
  "type-aliases/ConditionalType.md",
  "type-aliases/ExternalReferenceType.md",
  "type-aliases/FunctionType.md",
  "type-aliases/IndexAccessType.md",
  "type-aliases/IntersectionType.md",
  "type-aliases/LiteralType.md",
  "type-aliases/PartialMappedType.md",
  "type-aliases/PrimitiveType.md",
  "type-aliases/ReadonlyMappedType.md",
  "type-aliases/StringLiteralType.md",
  "type-aliases/TupleType.md",
  "type-aliases/TypeWithExternalSymbolLinkMapping.md",
  "type-aliases/TypeWithTypeParams.md",
  "type-aliases/UnionType.md",
  "variables/objectLiteralVariable.md",
  "variables/stringVariable.md",
  "variables/typeOperatorVariable.md",
]
`;

exports[`Urls should gets Urls for single entry points: outputFileStrategy: modules 1`] = `
[
  "README.md",
  "exports.md",
]
`;

exports[`Urls should gets Urls for single entry points: outputFileStrategy: modules 2`] = `
[
  "index.md",
]
`;
